File: Src/ECS/Entity/Core/EntityRelationshipManager.cs
using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// Entity 关系管理器
/// 负责管理 Entity 之间的关系，采用三索引结构实现高效查询
/// </summary>
public static class EntityRelationshipManager
{
    private static readonly Log _log = new("EntityRelationshipManager");

    // ==================== 数据结构 ====================

    /// <summary>
    /// 关系记录（公开以支持查询方法返回）
    /// </summary>
    public class RelationshipRecord
    {
        public string ParentEntityId { get; set; } = string.Empty;
        public string ChildEntityId { get; set; } = string.Empty;
        public string RelationType { get; set; } = string.Empty;
        public Dictionary<string, object> Data { get; set; } = new();
    }

    // 主存储：relationshipId -> RelationshipRecord
    private static readonly Dictionary<string, RelationshipRecord> _relationships = new();

    // 三索引结构
    private static readonly Dictionary<string, HashSet<string>> _parentIndex = new();  // parentEntityId -> Set<relationshipId>
    private static readonly Dictionary<string, HashSet<string>> _childIndex = new();   // childEntityId -> Set<relationshipId>
    private static readonly Dictionary<string, HashSet<string>> _typeIndex = new();    // relationType -> Set<relationshipId>

    // ==================== 核心方法 ====================

    /// <summary>
    /// 生成关系 ID
    /// </summary>
    private static string GenerateRelationshipId(string parentId, string childId, string type)
        => $"{parentId}:{childId}:{type}";

    /// <summary>
    /// 添加关系
    /// </summary>
    /// <param name="parentId">父 Entity ID</param>
    /// <param name="childId">子 Entity ID</param>
    /// <param name="relationType">关系类型</param>
    /// <param name="data">关系附加数据（可选）</param>
    /// <returns>是否成功添加</returns>
    public static bool AddRelationship(string parentId, string childId, string relationType, Dictionary<string, object>? data = null)
    {
        string relationshipId = GenerateRelationshipId(parentId, childId, relationType);

        // 检查重复
        if (_relationships.ContainsKey(relationshipId))
        {
            _log.Warn($"关系已存在: {parentId} -> {childId} ({relationType})");
            return false;
        }

        // 子 Entity 的关系不能重复（一个物品只能属于一个玩家）
        if (GetParentEntitiesByChildAndType(childId, relationType).Any())
        {
            _log.Warn($"子 Entity 已存在关系: {childId} ({relationType})");
            return false;
        }

        // 创建记录
        var record = new RelationshipRecord
        {
            ParentEntityId = parentId,
            ChildEntityId = childId,
            RelationType = relationType,
            Data = data ?? new()
        };

        // 添加到主存储
        _relationships[relationshipId] = record;

        // 更新三索引
        GetOrCreateSet(_parentIndex, parentId).Add(relationshipId);
        GetOrCreateSet(_childIndex, childId).Add(relationshipId);
        GetOrCreateSet(_typeIndex, relationType).Add(relationshipId);

        _log.Debug($"已添加关系: {parentId} -> {childId} ({relationType})");
        return true;
    }

    /// <summary>
    /// 移除关系
    /// </summary>
    public static bool RemoveRelationship(string parentId, string childId, string relationType)
    {
        string relationshipId = GenerateRelationshipId(parentId, childId, relationType);

        if (!_relationships.Remove(relationshipId))
        {
            _log.Warn($"关系不存在: {parentId} -> {childId} ({relationType})");
            return false;
        }

        // 从三索引移除
        _parentIndex.GetValueOrDefault(parentId)?.Remove(relationshipId);
        _childIndex.GetValueOrDefault(childId)?.Remove(relationshipId);
        _typeIndex.GetValueOrDefault(relationType)?.Remove(relationshipId);

        // 清理空集合
        CleanupEmptySet(_parentIndex, parentId);
        CleanupEmptySet(_childIndex, childId);
        CleanupEmptySet(_typeIndex, relationType);

        _log.Debug($"已移除关系: {parentId} -> {childId} ({relationType})");
        return true;
    }

    /// <summary>
    /// 设置关系数据
    /// </summary>
    public static bool SetRelationshipData(string parentId, string childId, string relationType, Dictionary<string, object> data)
    {
        string relationshipId = GenerateRelationshipId(parentId, childId, relationType);

        if (!_relationships.TryGetValue(relationshipId, out var record))
        {
            _log.Warn($"关系不存在: {parentId} -> {childId} ({relationType})");
            return false;
        }

        // 合并数据
        foreach (var kvp in data)
        {
            record.Data[kvp.Key] = kvp.Value;
        }

        return true;
    }

    /// <summary>
    /// 获取关系数据
    /// </summary>
    public static Dictionary<string, object>? GetRelationshipData(string parentId, string childId, string relationType)
    {
        string relationshipId = GenerateRelationshipId(parentId, childId, relationType);
        return _relationships.GetValueOrDefault(relationshipId)?.Data;
    }

    /// <summary>
    /// 检查关系是否存在
    /// </summary>
    public static bool HasRelationship(string parentId, string childId, string relationType)
    {
        string relationshipId = GenerateRelationshipId(parentId, childId, relationType);
        return _relationships.ContainsKey(relationshipId);
    }

    // ==================== 查询接口 ====================

    /// <summary>
    /// 获取父 Entity 的所有子 Entity（指定关系类型）
    /// 常用场景：获取玩家的所有物品
    /// </summary>
    public static IEnumerable<string> GetChildEntitiesByParentAndType(string parentId, string relationType)
    {
        if (!_parentIndex.TryGetValue(parentId, out var relationshipIds))
            return Enumerable.Empty<string>();

        return relationshipIds
            .Select(id => _relationships.GetValueOrDefault(id))
            .Where(r => r?.RelationType == relationType)
            .Select(r => r!.ChildEntityId);
    }

    /// <summary>
    /// 获取子 Entity 的所有父 Entity（指定关系类型）
    /// 常用场景：获取物品的拥有者
    /// </summary>
    public static IEnumerable<string> GetParentEntitiesByChildAndType(string childId, string relationType)
    {
        if (!_childIndex.TryGetValue(childId, out var relationshipIds))
            return Enumerable.Empty<string>();

        return relationshipIds
            .Select(id => _relationships.GetValueOrDefault(id))
            .Where(r => r?.RelationType == relationType)
            .Select(r => r!.ParentEntityId);
    }

    /// <summary>
    /// 获取父 Entity 的所有关系
    /// </summary>
    public static IEnumerable<RelationshipRecord> GetRelationshipsByParent(string parentId)
    {
        if (!_parentIndex.TryGetValue(parentId, out var relationshipIds))
            return Enumerable.Empty<RelationshipRecord>();

        return relationshipIds
            .Select(id => _relationships.GetValueOrDefault(id))
            .Where(r => r != null)!;
    }

    /// <summary>
    /// 获取子 Entity 的所有关系
    /// </summary>
    public static IEnumerable<RelationshipRecord> GetRelationshipsByChild(string childId)
    {
        if (!_childIndex.TryGetValue(childId, out var relationshipIds))
            return Enumerable.Empty<RelationshipRecord>();

        return relationshipIds
            .Select(id => _relationships.GetValueOrDefault(id))
            .Where(r => r != null)!;
    }

    /// <summary>
    /// 获取指定类型的所有关系
    /// </summary>
    public static IEnumerable<RelationshipRecord> GetRelationshipsByType(string relationType)
    {
        if (!_typeIndex.TryGetValue(relationType, out var relationshipIds))
            return Enumerable.Empty<RelationshipRecord>();

        return relationshipIds
            .Select(id => _relationships.GetValueOrDefault(id))
            .Where(r => r != null)!;
    }

    // ==================== 批量操作 ====================

    /// <summary>
    /// 移除 Entity 的所有关系（Entity 销毁时调用）
    /// </summary>
    public static void RemoveAllRelationships(string entityId)
    {
        // 清理作为父 Entity 的关系
        if (_parentIndex.TryGetValue(entityId, out var parentRels))
        {
            foreach (var relId in parentRels.ToList())
            {
                var record = _relationships.GetValueOrDefault(relId);
                if (record != null)
                    RemoveRelationship(record.ParentEntityId, record.ChildEntityId, record.RelationType);
            }
        }

        // 清理作为子 Entity 的关系
        if (_childIndex.TryGetValue(entityId, out var childRels))
        {
            foreach (var relId in childRels.ToList())
            {
                var record = _relationships.GetValueOrDefault(relId);
                if (record != null)
                    RemoveRelationship(record.ParentEntityId, record.ChildEntityId, record.RelationType);
            }
        }

        _log.Debug($"已清理 Entity 的所有关系: {entityId}");
    }

    /// <summary>
    /// 移除指定类型的所有关系
    /// </summary>
    public static void RemoveRelationshipsByType(string relationType)
    {
        if (!_typeIndex.TryGetValue(relationType, out var relationshipIds))
            return;

        // 复制列表避免迭代时修改
        var idsToRemove = relationshipIds.ToList();

        foreach (var id in idsToRemove)
        {
            var record = _relationships.GetValueOrDefault(id);
            if (record != null)
                RemoveRelationship(record.ParentEntityId, record.ChildEntityId, record.RelationType);
        }

        _log.Info($"已移除所有类型为 {relationType} 的关系，共 {idsToRemove.Count} 个");
    }

    // ==================== 工具方法 ====================

    /// <summary>
    /// 获取或创建 HashSet
    /// </summary>
    private static HashSet<string> GetOrCreateSet(Dictionary<string, HashSet<string>> dict, string key)
    {
        if (!dict.TryGetValue(key, out var set))
        {
            set = new HashSet<string>();
            dict[key] = set;
        }
        return set;
    }

    /// <summary>
    /// 清理空集合
    /// </summary>
    private static void CleanupEmptySet(Dictionary<string, HashSet<string>> dict, string key)
    {
        if (dict.TryGetValue(key, out var set) && set.Count == 0)
        {
            dict.Remove(key);
        }
    }

    /// <summary>
    /// 清理所有数据
    /// </summary>
    public static void Clear()
    {
        _relationships.Clear();
        _parentIndex.Clear();
        _childIndex.Clear();
        _typeIndex.Clear();
        _log.Info("EntityRelationshipManager 已清空");
    }

    /// <summary>
    /// 获取统计信息
    /// </summary>
    public static (int TotalRelationships, int ParentCount, int ChildCount, int TypeCount) GetStats()
    {
        return (
            _relationships.Count,
            _parentIndex.Count,
            _childIndex.Count,
            _typeIndex.Count
        );
    }
}

File: 参考/ECS/Entity/EntityRelationshipManager.ts
/** @noSelfInFile **/

import { Logger } from "../../base/object/工具/logger";

const logger = Logger.createLogger("EntityRelationshipManager");

/**
 * Entity关系数据接口
 */
interface RelationshipData {
  [key: string]: any;
}

/**
 * 关系记录接口
 */
interface RelationshipRecord {
  /** 父Entity ID */
  parentEntityId?: string;
  /** 子Entity ID */
  childEntityId?: string;
  /** 关系类型 */
  relationType?: string;
  /** 关系附加数据 */
  data?: RelationshipData;
}

/**
 * Entity关系管理器
 * 
 * 负责管理Entity之间的关系，这是现代ECS架构中的重要组成部分
 * 使用单层Map优化性能和维护性
 */
export class EntityRelationshipManager {
  // 单例模式
  private static instance: EntityRelationshipManager;

  // 关系存储：relationshipId -> RelationshipRecord
  private relationshipsMap: Map<string, RelationshipRecord> = new Map();

  // 父Entity索引：parentEntityId -> Set<relationshipId>
  private parentIndexMap: Map<string, Set<string>> = new Map();

  // 子Entity索引：childEntityId -> Set<relationshipId>
  private childIndexMap: Map<string, Set<string>> = new Map();

  // 关系类型索引：relationType -> Set<relationshipId>
  private typeIndexMap: Map<string, Set<string>> = new Map();

  private constructor() {
    // logger.debug("EntityRelationshipManager created");
  }

  public static getInstance(): EntityRelationshipManager {
    if (!EntityRelationshipManager.instance) {
      EntityRelationshipManager.instance = new EntityRelationshipManager();
    }
    return EntityRelationshipManager.instance;
  }

  // ====================== 关系管理 ======================
  /**
   * 生成关系ID
   */
  private generateRelationshipId(parentEntityId: string, childEntityId: string, relationType: string): string {
    return `${parentEntityId}:${childEntityId}:${relationType}`;
  }

  /**
   * 添加Entity关系【增】
   * @param parentEntityId 父EntityId
   * @param childEntityId 子EntityId
   * @param relationType 关系类型，可选，如果不提供则根据两个Entity的类型自动生成
   * @param data 关系数据
   * @returns 是否成功添加
   */
  addRelationship(parentEntityId: string, childEntityId: string, relationType?: string, data?: RelationshipData): boolean {
    if (!parentEntityId || !childEntityId) {
      logger.warn("无效的关系参数：缺少Entity ID");
      return false;
    }

    // 如果没有提供关系类型，则自动生成
    if (!relationType) {
      relationType = `${parentEntityId}_${childEntityId}`;
    }

    const relationshipId = this.generateRelationshipId(parentEntityId, childEntityId, relationType);

    // 检查关系是否已存在
    if (this.relationshipsMap.has(relationshipId)) {
      logger.warn(`关系已存在: ${parentEntityId} -> ${childEntityId} (${relationType})`);
      return false;
    }
    // 子Entity的关系不能重复，父Entity的关系可以重复，比如父Entity是玩家，子Entity是单位，一个玩家可以有多个单位，但是一个单位只能属于一个玩家
    if (this.getRelationshipsByChildAndType(childEntityId, relationType).length > 0) {
      logger.warn(`子Entity已存在关系: ${childEntityId} (${relationType})`);
      return false;
    }

    const now = Date.now();
    const record: RelationshipRecord = {
      parentEntityId,
      childEntityId,
      relationType,
      data: data || {},
    };

    // 添加到主存储
    this.relationshipsMap.set(relationshipId, record);

    // 更新父Entity索引
    if (!this.parentIndexMap.has(parentEntityId)) {
      this.parentIndexMap.set(parentEntityId, new Set());
    }
    this.parentIndexMap.get(parentEntityId)!.add(relationshipId);

    // 更新子Entity索引
    if (!this.childIndexMap.has(childEntityId)) {
      this.childIndexMap.set(childEntityId, new Set());
    }
    this.childIndexMap.get(childEntityId)!.add(relationshipId);

    // 更新类型索引
    if (!this.typeIndexMap.has(relationType)) {
      this.typeIndexMap.set(relationType, new Set());
    }
    this.typeIndexMap.get(relationType)!.add(relationshipId);

    logger.debug(`已添加关系: ${parentEntityId} -> ${childEntityId} (${relationType})`);
    return true;
  }

  /**
   * 移除关系【删】
   * @param parentEntityId 父EntityId
   * @param childEntityId 子EntityId
   * @param relationType 关系类型
   * @returns 是否成功移除
   */
  removeRelationship(parentEntityId: string, childEntityId: string, relationType: string): boolean {
    if (!parentEntityId || !childEntityId || !relationType) {
      logger.warn("无效的关系参数");
      return false;
    }

    const relationshipId = this.generateRelationshipId(parentEntityId, childEntityId, relationType);

    // 检查关系是否存在
    if (!this.relationshipsMap.has(relationshipId)) {
      logger.warn(`关系不存在: ${parentEntityId} -> ${childEntityId} (${relationType})`);
      return false;
    }

    // 从主存储移除
    this.relationshipsMap.delete(relationshipId);

    // 从父Entity索引移除
    const parentSet = this.parentIndexMap.get(parentEntityId);
    if (parentSet) {
      parentSet.delete(relationshipId);
      if (parentSet.size === 0) {
        this.parentIndexMap.delete(parentEntityId);
      }
    }

    // 从子Entity索引移除
    const childSet = this.childIndexMap.get(childEntityId);
    if (childSet) {
      childSet.delete(relationshipId);
      if (childSet.size === 0) {
        this.childIndexMap.delete(childEntityId);
      }
    }

    // 从类型索引移除
    const typeSet = this.typeIndexMap.get(relationType);
    if (typeSet) {
      typeSet.delete(relationshipId);
      if (typeSet.size === 0) {
        this.typeIndexMap.delete(relationType);
      }
    }

    logger.debug(`已移除关系: ${parentEntityId} -> ${childEntityId} (${relationType})`);
    return true;
  }

  /**
   * 设置关系数据【改】
   * @param parentEntityId 父EntityId
   * @param childEntityId 子EntityId
   * @param relationType 关系类型
   * @param data 关系数据
   * @returns 是否成功设置
   */
  setRelationshipData(parentEntityId: string, childEntityId: string, relationType: string, data: RelationshipData): boolean {
    const relationshipId = this.generateRelationshipId(parentEntityId, childEntityId, relationType);
    const record = this.relationshipsMap.get(relationshipId);
    if (!record) {
      logger.warn(`关系不存在: ${parentEntityId} -> ${childEntityId} (${relationType})`);
      return false;
    }
    record.data = { ...record.data, ...data };

    logger.debug(`已更新关系数据: ${parentEntityId} -> ${childEntityId} (${relationType})`);
    return true;
  }

  /**
   * 检查关系是否存在【查】
   * @param parentEntityId 父EntityId
   * @param childEntityId 子EntityId
   * @param relationType 关系类型
   * @returns 是否存在
   */
  hasRelationship(parentEntityId: string, childEntityId: string, relationType: string): boolean {
    const relationshipId = this.generateRelationshipId(parentEntityId, childEntityId, relationType);
    const has = this.relationshipsMap.has(relationshipId);
    if (!has) {
      logger.warn(`关系不存在: ${parentEntityId} -> ${childEntityId} (${relationType})`);
    }
    return has;
  }

  /**
   * 获取指定关系数据【查】
   * @param parentEntityId 父EntityId
   * @param childEntityId 子EntityId
   * @param relationType 关系类型 
   * @returns 关系数据
   */
  getRelationshipData(parentEntityId: string, childEntityId: string, relationType: string): RelationshipData | null {
    const relationshipId = this.generateRelationshipId(parentEntityId, childEntityId, relationType);
    const record = this.relationshipsMap.get(relationshipId);
    return record ? record.data || {} : null;
  }

  /**
   * 根据关系ID集合获取关系记录列表
   * @param relationshipIds 关系ID集合
   * @returns 关系记录列表
   */
  getRelationshipRecords(relationshipIds: Set<string>): RelationshipRecord[] {
    const records: RelationshipRecord[] = [];
    for (const id of relationshipIds) {
      const record = this.relationshipsMap.get(id);
      if (record) {
        records.push(record);
      }
    }
    return records;
  }
  /**
   * 获取父Entity的所有关系
   */
  getRelationshipsByParent(parentEntityId: string): RelationshipRecord[] {
    const relationshipIds = this.parentIndexMap.get(parentEntityId);
    if (!relationshipIds) return [];
    return this.getRelationshipRecords(relationshipIds);
  }

  /**
   * 获取子Entity的所有关系
   * @param childEntityId 子Entity的ID
   */
  getRelationshipsByChild(childEntityId: string): RelationshipRecord[] {
    const relationshipIds = this.childIndexMap.get(childEntityId);
    if (!relationshipIds) return [];
    return this.getRelationshipRecords(relationshipIds);
  }

  /**
   * 获取指定类型的所有关系
   * @param relationType 关系类型
   */
  getRelationshipsByType(relationType: string): RelationshipRecord[] {
    const relationshipIds = this.typeIndexMap.get(relationType);
    if (!relationshipIds) return [];
    return this.getRelationshipRecords(relationshipIds);
  }

  /**
   * 获取父Entity的所有相关Entity（指定关系类型），常用，比如获取UnitEntity的多个物品ItemEntity
   * @param parentEntityId 父EntityId
   * @param relationType 关系类型
   * @returns 子EntityId列表
   */
  getRelationshipsByParentAndType(parentEntityId: string, relationType: string): string[] {
    const relationshipIds = this.parentIndexMap.get(parentEntityId);
    if (!relationshipIds) return [];

    const childEntityIds: string[] = [];
    for (const id of relationshipIds) {
      const record = this.relationshipsMap.get(id);
      if (record && record.relationType === relationType && record.childEntityId) {
        childEntityIds.push(record.childEntityId);
      }
    }
    return childEntityIds;
  }

  /**
   * 获取子Entity的所有相关Entity（指定关系类型），常用，比如获取ItemEntity的所有拥有者UnitEntity
   * @param childEntityId 子EntityId
   * @param relationType 关系类型
   * @returns 父EntityId列表
   */
  getRelationshipsByChildAndType(childEntityId: string, relationType: string): string[] {
    const relationshipIds = this.childIndexMap.get(childEntityId);
    if (!relationshipIds) return [];

    const parentEntityIds: string[] = [];
    for (const id of relationshipIds) {
      const record = this.relationshipsMap.get(id);
      if (record && record.relationType === relationType && record.parentEntityId) {
        parentEntityIds.push(record.parentEntityId);
      }
    }
    return parentEntityIds;
  }

  // ====================== 批量操作 ======================

  /**
   * 移除指定类型的所有关系
   */
  removeRelationshipsByType(relationType: string): void {
    const relationshipIds = this.typeIndexMap.get(relationType);
    if (!relationshipIds) return;

    // 复制ID列表，避免在迭代过程中修改
    const idsToRemove = [...relationshipIds];

    for (const id of idsToRemove) {
      const record = this.relationshipsMap.get(id);
      if (record && record.parentEntityId && record.childEntityId && record.relationType) {
        this.removeRelationship(record.parentEntityId, record.childEntityId, record.relationType);
      }
    }

    logger.debug(`已移除所有类型为 ${relationType} 的关系`);
  }

  /**
   * 移除Entity的所有关系，Entity执行destroy时运行，将与自己有关的关系全部销毁
   * @param entityId Entity.getId()
   */
  removeAllRelationships(entityId: string): void {
    // 移除作为父Entity的所有关系
    const parentRelationshipIds = this.parentIndexMap.get(entityId);
    if (parentRelationshipIds) {
      const idsToRemove = [...parentRelationshipIds];
      for (const id of idsToRemove) {
        const record = this.relationshipsMap.get(id);
        if (record && record.parentEntityId && record.childEntityId && record.relationType) {
          this.removeRelationship(record.parentEntityId, record.childEntityId, record.relationType);
        }
      }
    }

    // 移除作为子Entity的所有关系
    const childRelationshipIds = this.childIndexMap.get(entityId);
    if (childRelationshipIds) {
      const idsToRemove = [...childRelationshipIds];
      for (const id of idsToRemove) {
        const record = this.relationshipsMap.get(id);
        if (record && record.parentEntityId && record.childEntityId && record.relationType) {
          this.removeRelationship(record.parentEntityId, record.childEntityId, record.relationType);
        }
      }
    }

    logger.debug(`已移除Entity的所有关系: ${entityId}`);
  }

  // ====================== 工具方法 ======================
  /**
   * 清理所有数据
   */
  clear(): void {
    this.relationshipsMap.clear();
    this.parentIndexMap.clear();
    this.childIndexMap.clear();
    this.typeIndexMap.clear();

    logger.info("EntityRelationshipManager 已清理所有数据");
  }

  /**
   * 销毁管理器
   */
  destroy(): void {
    this.clear();
    logger.info("EntityRelationshipManager 已销毁");
  }
}